--- cups-pdf-3.0beta2/src/cups-pdf.c.desktop	2016-05-09 11:50:02.355531718 +0200
+++ cups-pdf-3.0beta2/src/cups-pdf.c	2016-05-09 11:53:05.164945770 +0200
@@ -438,6 +438,45 @@ static void announce_printers() {
   return;
 }
 
+static char *getdesktopdir(struct passwd *passwd) {
+	static char bufret[BUFSIZE];
+	char *need, *ret, buftmp[BUFSIZE];
+	FILE *fic;
+
+	// Default (if not found) = $HOME
+	ret=passwd->pw_dir;
+
+	snprintf(buftmp, BUFSIZE, "%s/.config/user-dirs.dirs", passwd->pw_dir);
+	fic = fopen(buftmp, "r");
+	if (fic) {
+		while (fgets(buftmp, BUFSIZE, fic)) {
+			need=strstr(buftmp, "XDG_DESKTOP_DIR=");
+			if (need) {
+				need+=16;
+				// Remove opening quote
+				if ((*need)=='"') need++;
+				if (memcmp(need, "$HOME", 5)) {
+					strncpy(bufret, need, BUFSIZE);
+				}
+				else {
+					snprintf(bufret, BUFSIZE, "%s%s", passwd->pw_dir, need+5);
+				}
+				// Remove closing quote
+				need=strrchr(bufret, '"');
+				if (need) *need=0;
+				// Return internal buffer
+				ret = bufret;
+				
+			        log_event(CPDEBUG, "XDG_DESKTOP_DIR:", ret);
+			}
+		} 
+		fclose(fic);
+	}
+	else
+	        log_event(CPERROR, "Can't read", buftmp);
+	return ret;
+}
+
 static char *preparedirname(struct passwd *passwd, char *uname) {
   int size;
   char bufin[BUFSIZE], bufout[BUFSIZE], *needle, *cptr;
@@ -447,6 +486,16 @@ static char *preparedirname(struct passw
     uname=uname+size;
 
   strncpy(bufin, Conf_Out, BUFSIZE);
+
+  do {
+    needle=strstr(bufin, "${DESKTOP}");
+    if (needle == NULL)
+      break;
+    needle[0]='\0';
+    cptr=needle+10;
+    snprintf(bufout, BUFSIZE, "%s%s%s", bufin, getdesktopdir(passwd), cptr);
+    strncpy(bufin, bufout, BUFSIZE);
+  } while (needle != NULL);
   do {
     needle=strstr(bufin, "${HOME}");
     if (needle == NULL)
